src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,171 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final int INDEX_BY_START_REMOVING = 0;
+    private static final int FLAG_THAT_VALUE_IS_UNREAL = -1;
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            head = new Node<>(null, value,null);
+            tail = head;
+        } else {
+            Node<T> currentNode = new Node<>(tail, value, null);
+            tail.next = currentNode;
+            tail = currentNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index + " is not valid");
+        } else {
+            if (index == size) {
+                add(value);
+                return;
+            }
+            if (index == 0) {
+                Node<T> newNode = new Node<>(null, value, head);
+                head.prev = newNode;
+                head = newNode;
+                size++;
+                return;
+            }
+            Node<T> currentNode = getNodeByIndex(index);
+            Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T temp = getNodeByIndex(index).item;
+        getNodeByIndex(index).item = value;
+        return temp;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = getNodeByIndex(index);
+        unLink(currentNode);
+        size--;
+        return currentNode.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        int index = INDEX_BY_START_REMOVING;
+        while (!compare(currentNode.item, object)) {
+            if (index == size - 1) {
+                index = FLAG_THAT_VALUE_IS_UNREAL;
+                break;
+            }
+            currentNode = currentNode.next;
+            index++;
+        }
+        if (index != FLAG_THAT_VALUE_IS_UNREAL) {
+            unLink(currentNode);
+            size--;
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unLink(Node<T> currentNode) {
```

Comment:
let's set item to null
but before calling this method - please save the old item where you need it

Reply:
you can optimize code inside this method using if-else and add size decrement at the end